File: server/actions/vpn_down.go
package actions

import (
	"vpn-sandbox/core"
	"vpn-sandbox/utils"
)

func VpnDown() {
	utils.LogLn("VpnDown: Entry")

	// restore resolv.conf
	utils.LogLn("Restoring resolv.conf")
	utils.RestoreResolvConf()

	// get host gateway from resolv.conf
	hostGateway := utils.GetHostGateway()
	utils.LogLn("host gateway: " + hostGateway)

	if core.Testing {
		utils.PublishEvent(utils.Event{Name: "vpn-down", Context: map[string]interface{}{}})
		utils.LogLn("Skipping vpn down actions for testing")
		return
	}

	// restore routes
	utils.LogLn("Restoring routes")

	// Remove all existing default routes
	utils.RunCommand(false, "/sbin/ip", "route", "del", "default")

	// Add default gateway
	utils.RunCommand(false, "/sbin/ip", "route", "add", "default", "via", hostGateway, "dev", "eth0")

	// Set firewall rules
	// Flush existing rules to start fresh
	utils.RunCommand(false, "/sbin/iptables", "-F")

	// Allow related and established connections (for existing sessions to work)
	utils.RunCommand(false, "/sbin/iptables", "-A", "INPUT", "-m", "conntrack", "--ctstate", "RELATED,ESTABLISHED", "-j", "ACCEPT")

	// Allow incoming connections only on port 80
	utils.RunCommand(false, "/sbin/iptables", "-A", "INPUT", "-p", "tcp", "--dport", "80", "-j", "ACCEPT")

	// Drop all other incoming connections
	utils.RunCommand(false, "/sbin/iptables", "-A", "INPUT", "-j", "DROP")

	// Trigger vpn down actions
	utils.LogLn("Triggering vpn down actions")
	utils.PublishEvent(utils.Event{Name: "vpn-down", Context: map[string]interface{}{}})

	// Run app script
	utils.LogLn("Stopping app script")
	utils.RunCommand(false, core.AppScript, "down")
}

File: server/actions/vpn_down.py
from utils import log_ln, run_command, publish_event_sync, Event, restore_resolv_conf
import core
from .apps import run_legacy_app_script, run_managed_apps_action

def vpn_down():
    log_ln("VpnDown: Entry")
    
    if not core.Testing:
        log_ln("Stopping app lifecycle hooks")
        run_legacy_app_script("down")
        run_managed_apps_action("down")
            
    log_ln("Triggering vpn down actions")
    publish_event_sync(Event("vpn-down", {}))
    
    log_ln("Restoring resolv.conf")
    restore_resolv_conf()
    
    log_ln("host gateway: " + core.HostGateway)
    
    if core.Testing:
        log_ln("Skipping vpn down actions for testing")
        return
        
    log_ln("Restoring routes")
    run_command(False, "/sbin/ip", "route", "del", "default")
    run_command(False, "/sbin/ip", "route", "add", "default", "via", core.HostGateway, "dev", "eth0")
    
    run_command(False, "/usr/sbin/iptables", "-F")
    run_command(False, "/usr/sbin/ip6tables", "-F")
    
    run_command(False, "/usr/sbin/iptables", "-A", "INPUT", "-m", "conntrack", "--ctstate", "RELATED,ESTABLISHED", "-j", "ACCEPT")
    run_command(False, "/usr/sbin/iptables", "-A", "INPUT", "-p", "tcp", "--dport", "80", "-j", "ACCEPT")
    run_command(False, "/usr/sbin/iptables", "-A", "INPUT", "-j", "DROP")
